ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateArchive.java

```diff
@@ -47,59 +47,67 @@
 
   public BonsaiWorldStateArchive(final StorageProvider provider, final Blockchain blockchain) {
     this.blockchain = blockchain;
 
     worldStateStorage = new BonsaiWorldStateKeyValueStorage(provider);
     persistedState = new BonsaiPersistedWorldState(this, worldStateStorage);
-    layeredWorldStates = new HashMap<>();
+    layeredWorldStatesByHash = new HashMap<>();
   }
 
   @Override
   public Optional<WorldState> get(final Hash rootHash, final Hash blockHash) {
-    if (layeredWorldStates.containsKey(rootHash)) {
-      return Optional.of(layeredWorldStates.get(blockHash));
+    if (layeredWorldStatesByHash.containsKey(rootHash)) {
+      return Optional.of(layeredWorldStatesByHash.get(blockHash));
     } else if (rootHash.equals(persistedState.blockHash())) {
       return Optional.of(persistedState);
     } else {
       return Optional.empty();
     }
   }
 
-  void addLayeredWorldState(final BonsaiLayeredWorldState worldState) {
-    layeredWorldStates.put(worldState.blockHash(), worldState);
+  public void addLayeredWorldState(
+      final BonsaiWorldView persistedWorldState,
+      final long blockNumber,
+      final Hash worldStateRootHash,
+      final TrieLogLayer trieLog) {
+    final BonsaiLayeredWorldState bonsaiLayeredWorldState =
+        new BonsaiLayeredWorldState(
+            Optional.of(persistedWorldState), blockNumber, worldStateRootHash, trieLog);
+    layeredWorldStatesByHash.put(bonsaiLayeredWorldState.blockHash(), bonsaiLayeredWorldState);
+    if (blockNumber > 0) {
+      final Optional<Hash> blockHashByNumber = blockchain.getBlockHashByNumber(blockNumber - 1);
+      if (blockHashByNumber.isPresent()
+          && layeredWorldStatesByHash.containsKey(blockHashByNumber.get())) {
+        layeredWorldStatesByHash
+            .get(blockHashByNumber.get())
+            .setParent(Optional.of(bonsaiLayeredWorldState));
+      }
+    }
   }
 
   public Optional<TrieLogLayer> getTrieLogLayer(final Hash blockHash) {
-    if (layeredWorldStates.containsKey(blockHash)) {
-      return Optional.of(layeredWorldStates.get(blockHash).getTrieLog());
+    if (layeredWorldStatesByHash.containsKey(blockHash)) {
+      return Optional.of(layeredWorldStatesByHash.get(blockHash).getTrieLog());
     } else {
       return worldStateStorage.getTrieLog(blockHash).map(TrieLogLayer::fromBytes);
     }
   }
 
   @Override
   public boolean isWorldStateAvailable(final Hash rootHash, final Hash blockHash) {
-    return layeredWorldStates.containsKey(blockHash)
+    return layeredWorldStatesByHash.containsKey(blockHash)
         || persistedState.blockHash().equals(blockHash)
         || worldStateStorage.isWorldStateAvailable(rootHash, blockHash);
   }
 
   @Override
   public Optional<MutableWorldState> getWorldState(final Hash rootHash, final Hash blockHash) {
-    if (layeredWorldStates.containsKey(blockHash)) {
-      return Optional.of(layeredWorldStates.get(blockHash));
-    } else {
```

Comment:
If we don't load TrieLog layered world states from disk how do we then do traces beyond near chain head?  We won't be able to do eth_call 1MM blocks back but we should be able to do trace_* from 1MM blocks back if we have the TrieLogs.

Reply:
Indeed I just put back the loading of the layer from the disk but in this case I put the nextLayer to empty and I load it on demand (call getNextWorldView) so as not to load all layers for nothing